src/base/asyncfilestorage.cpp

```diff
@@ -0,0 +1,113 @@
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "asyncfilestorage.h"
+
+#include <QDebug>
+#include <QMetaObject>
+#include <QMutexLocker>
+#include <QSaveFile>
+#include <QTimer>
+
+AsyncFileStorage::AsyncFileStorage(const QDir &storageDir, QFile *lockFile, QObject *parent)
+    : QObject(parent)
+    , m_storageDir(storageDir)
+    , m_lockFile(lockFile)
+{
+}
+
+AsyncFileStorage::~AsyncFileStorage()
+{
+    m_lockFile->close();
+    m_lockFile->remove();
+    delete m_lockFile;
+}
+
+AsyncFileStorage *AsyncFileStorage::create(const QString &storageFolder, QObject *parent)
+{
+    QDir storageDir(storageFolder);
+    QFile *lockFile = new QFile(storageDir.absoluteFilePath(QStringLiteral("storage.lock")));
+    if (storageDir.exists() || storageDir.mkpath(storageDir.absolutePath())) {
+        if (lockFile->open(QFile::WriteOnly))
+            return new AsyncFileStorage(storageDir, lockFile, parent);
+    }
+
+    delete lockFile;
+    return nullptr;
+}
+
+void AsyncFileStorage::store(const QString &fileName, const QByteArray &data)
+{
+    QMutexLocker locker(&m_jobsMutex);
+    m_jobs.insert(fileName, data);
+    qDebug() << "AsyncFileStorage: Added job for" << fileName;
+    if (m_isIdle) {
+        m_isIdle = false;
+        // The function will be run in the thread of receiver (this AsyncFileStorage object).
+        QTimer::singleShot(0, this, &AsyncFileStorage::processPendingJobs);
+    }
+}
+
+void AsyncFileStorage::processPendingJobs()
+{
+    m_jobsMutex.lock();
+    QStringList fileNames = m_jobs.keys();
+    m_jobsMutex.unlock();
+
+    while (!fileNames.isEmpty()) {
+        foreach (const QString &fileName, fileNames) {
+            m_jobsMutex.lock();
+            const QByteArray data = m_jobs.take(fileName);
```

Comment:
But this class was designed as non interactive. It's mostly for writing data with short lifetime. Instructed and forgotten. Write error in this case is just a coincidence (which is very unlikely), or serious problems with the hardware, the OS (or brain) of the user. In the first case, it is sufficient to ignore this error and continue working (well, maybe to repeat the task if it has not yet been reset). In the second case, actually, the normal operation of the program impossible, but it is beyond the scope of this PR.